keps/sig-apps/0008-sidecarcontainer.md

```diff
@@ -0,0 +1,145 @@
+  - name: sidecar
+    image: sidecar-image
+    sidecar: true
+    command: ["do something to help my app"]
+
+```
+Sidecars will be started before normal containers but after init, so that they are ready before your main processes start.
+
+This will change the Pod startup to look like this:
+* Init containers start
+* Init containers finish
+* Sidecars start
+* Containers start
+
+During pod termination sidecars will be terminated last:
+* Containers sent SIGTERM
+* Sidecars sent SIGTERM
+
+Both will still obey the grace period before being sent a SIGKILL
+
+PreStop Hooks will be sent to sidecars and containers at the same time.
+This will be useful in scenarios such as when your sidecar is a proxy so that it knows to no longer accept inbound requests but can continue to allow outbound ones until the the primary containers have shut down.
+
+To solve the problem of Jobs that don't complete: When RestartPolicy!=Always if all normal containers have reached a terminal state (Succeeded for restartPolicy=OnFailure, or Succeeded/Failed for restartPolicy=Never), then all sidecar containers will be sent a SIGTERM.
+
+### Implementation Details/Notes/Constraints
+
+As this is a fairly large change I think it make sense to break this proposal down and phase in more functionality as we go, potential roadmap could look like:
+
+* Add sidecar field, use it for the shutdown triggering when RestartPolicy!=Always
+* Pre-stop hooks sent to sidecars before non sidecar containers
+* Sidecars are terminated after normal containers
+* Sidecars start before normal containers
+
+
+As this is a change to the Container spec we will be using feature gating, you will be required to explicitly enable this feature on the api server as recommended [here](https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#adding-unstable-features-to-stable-versions).
+
+### Risks and Mitigations
+
+You could set all containers to be `sidecar: true`, this seems wrong, so maybe the api should do a validation check that at least one container is not a sidecar.
+
+Init containers would be able to have `sidecar: true` applied to them as it's an additional field to the container spec, this doesn't currently make sense as init containers are ran sequentially. We could get around this by having the api throw a validation error if you try to use this field on an init container or just ignore the field.
+
+Having it as a boolean could cause problems later down the line if more lifecycle related flags were added, perhaps it makes more sense to have something like `lifecycle: Sidecar` to make it more future proof.
+
+## Graduation Criteria
+
+//TODO
+
+## Implementation History
+
+- 14th May 2018: Proposal Submitted
+
+
+## Alternatives
+
+One alternative would be to have a new field in the Pod Spec of `sidecarContainers:` where you could define a list of sidecar containers, however this would require more work in terms of updating tooling to support this.
```

Comment:
The `sidecarContainers` field has a certain symmetry with `initContainers` being separate from `containers` already. It also would introduce a nice layering, making it easier to reason about where to look for what:

* initContainers: stuff that runs to completion once to set up the Pod environment (local/ephemeral volumes, etc.)
* sidecarContainers: stuff that's not part of the main purpose of the Pod, but provides support services.
* containers: stuff that performs the main purpose of the Pod.

What kind of tooling updates are you referring to? Changes within the parts of k8s that ship together in a release? Or the ecosystem of 2nd and 3rd party tools?

If you're mostly concerned about the changes in the k8s core, I think we should prefer whatever choice makes the API cleaner in the long run, even if it's more work for us. Implementations can always be refactored, but interfaces are forever. Does anyone have examples of pain caused by adding `initContainers` as a separate field (not counting the mess caused by starting as an annotation)?

As for 2nd and 3rd party tools, one could argue it's a desirable feature that they would not be aware of the new containers in the separate field unless they have been updated to process it explicitly. Especially with sidecars that are injected at admission time by mutating webhooks, having extra stuff show up in `containers:` can be confusing and noisy for users and tooling. Since we're designing sidecars to semantically "wrap around" regular containers, keeping them invisible to "old clients" makes sense from the standpoint of encapsulation.

Reply:
I do like the parity with init containers if we went down the `sidecarContainers` route, it also gets around the problem of having it as a boolean flag. 

Just tooling in general, co-ordination within the ecosystem shouldn't be that bad (kube-state-metrics and kube dashboard spring to mind as tools that would need updating) but with third party tools it might cause some pain.
I'm just envisioning people upgrading to the new version of kube, moving their containers into the sidecars section and then various dashboards, graphs alerts etc getting very confused as the amount of containers in the pods decreases, containers they were monitoring are no longer visible etc. Maybe that pain is worth it for the long term benefit, I'm not sure.